spec/controllers/doorkeeper/authorizations_controller_spec.rb

```diff
@@ -89,10 +76,35 @@ def expect_successful_callback!
           expect(response.status).to redirect_to build_redirect_uri(error_params)
           expect(JSON.parse(response.body)).to eq(error_params)
         end
+
+        context 'when not logged in' do
+          let(:error_params) do
+            {
+              'error' => 'login_required',
+              'error_description' => 'The authorization server requires end-user authentication',
+              'state' => 'somestate'
+            }
+          end
+
+          it 'returns a login_required error' do
+            authorize! prompt: 'none', current_user: nil, state: 'somestate'
+
+            expect(response).to redirect_to build_redirect_uri(error_params)
+            expect(JSON.parse(response.body)).to eq(error_params)
+          end
+
+          it 'does not redirect to an invalid redirect_uri' do
+            authorize! prompt: 'none', current_user: nil, state: 'somestate', redirect_uri: 'https://evilapp.com'
+
+            expect(response).not_to be_redirect
+            expect(response.status).to eq 401
```

Comment:
See discussion in https://github.com/doorkeeper-gem/doorkeeper-openid_connect/issues/61#issuecomment-461883958:

> > @meagar: I think the correct behaviour here is to render a 422 error, rather than redirect the browser, regardless of the `prompt=none`.
>
> @toupeira: I'm not sure if the 422 error should have precedence, since we're already returning the OIDC/OAuth error.

Reply:
I've now changed the code so it checks `pre_auth.valid?` and prioritizes Doorkeeper's OAuth error response before falling back to the OIDC error.

This is currently also a 401, but see https://github.com/doorkeeper-gem/doorkeeper/issues/550 about maybe changing this to a 400 as suggested by the OAuth specs.